cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocProvisioner.java

```diff
@@ -152,6 +168,114 @@ public Cluster createCluster(ProvisionerContext context) throws Exception {
+   */
+  @Nullable
+  private Cluster tryReuseCluster(DataprocClient client, ProvisionerContext context, DataprocConf conf)
+    throws RetryableProvisionException, IOException {
+    if (!isReuseSupported(conf)) {
+      LOG.debug("Not checking cluter reuse, enabled: {}, skip delete: {}, idle ttl: {}, reuse threshold: {}",
+                conf.isClusterReuseEnabled(), conf.isSkipDelete(), conf.getIdleTTLMinutes(),
+                conf.getClusterReuseThresholdMinutes());
+      return null;
+    }
+
+    String defaultName = getClusterName(context);
+
+    //For idempotency, check if we already have the cluster allocated
+    Optional<Cluster> clusterOptional = getAllocatedCluster(defaultName, client);
+    if (clusterOptional.isPresent()) {
+      Cluster cluster = clusterOptional.get();
+      if (cluster.getStatus() == ClusterStatus.CREATING ||
+        cluster.getStatus() == ClusterStatus.RUNNING) {
+        LOG.debug("Found allocated cluster {}", cluster.getName());
+        return cluster;
+      } else {
+        LOG.debug("Preallocated cluster {} expired, will look for new one");
+        //Let's remove the reuse label to ensure new cluster will be picked up by getAllocatedCluster
+        try {
+          client.updateClusterLabels(cluster.getName(), Collections.emptyMap(), Collections.singleton(LABEL_REUSE_FOR));
+        } catch (Exception e) {
+          LOG.debug("Error during cluster cleanup, probably cluster died already", e);
+        }
+      }
+    }
+
+    Lock reuseLock = getSystemContext().getLock(REUSE_LOCK);
+    reuseLock.lock();
+    try {
+      Map<String, String> filter = new HashMap<>();
+      String normalizedProfileName = getNormalizedProfileName(context);
+      if (normalizedProfileName != null) {
+        filter.put(LABEL_PROFILE, normalizedProfileName);
+      }
+      filter.put(LABEL_VERSON, getVersionLabel());
+      filter.put(LABEL_REUSE_KEY, conf.getClusterReuseKey());
+      filter.put(LABEL_REUSE_UNTIL, "*");
+
+      Optional<Cluster> cluster = client.getClusters(ClusterStatus.RUNNING, filter, clientCluster -> {
+        //Verify reuse label
+        long reuseUntil = Long.valueOf(clientCluster.getLabelsOrDefault(LABEL_REUSE_UNTIL, "0"));
+        long now = System.currentTimeMillis();
+        if (reuseUntil < now) {
+          LOG.debug("Skipping expired cluster {}, reuse until {} before now {}",
+                    clientCluster.getClusterName(), reuseUntil, now);
+          return false;
+        }
+        return true;
+      }).findAny();
+
+      if (cluster.isPresent()) {
+        String clusterName = cluster.get().getName();
+        LOG.info("Found cluster to reuse: {}", clusterName);
+        // Add cdap-reuse-for to find cluster later if needed
+        // And remove reuseUntil to indicate the cluster is taken
+        client.updateClusterLabels(clusterName,
+                                   Collections.singletonMap(LABEL_REUSE_FOR, defaultName),
```

Comment:
I'm confused about what this label is for. How is defaultName different than the clusterName, and why is it removed on cluster delete? 

Would be good to add a javadoc that explains how all the labels are used.

Reply:
It's to find cluster later in io.cdap.cdap.runtime.spi.provisioner.dataproc.AbstractDataprocProvisioner#getRuntimeJobManager. Added a javadoc

Reply:
I don't understand why the label is needed when the cluster name is already given. Unless this label isn't the cluster name? If not, what is it?

Reply:
Don't see the javadoc, did the commit get pushed? Would be helpful to walk through an example of what the labels are the first run, then what they are the second run.

Reply:
I did some renames to clarify how it's used